Lookup plugin update: Update AWS SSM lookup to support getting parameters by path

mtb-xt opened the pull request:
##### SUMMARY
Update AWS SSM Parameter storage lookup plugin to support getting parameters by path

AWS SSM Parameter store lookup was added in PR https://github.com/ansible/ansible/pull/23460, but there were requests to add support to look up parameters by path. 

This is my attempt on that.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
plugins/lookup/ssm.py

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 3.6.2 (default, Jul 20 2017, 03:52:27) [GCC 7.1.1 20170630]
```

##### ADDITIONAL INFORMATION
Currently, the lookup plugin only gets a single parameter as in example below:
```
- debug: msg="{{ lookup('ssm', 'Hello', 'region=ap-southeast-2',  ) }}"
```

This PR adds these features:
You lookup the whole parameters hierarchy:
```
 - debug: msg="{{ lookup('ssm', '/Test', 'region=ap-southeast-2', 'bypath', 'recursive=true' ) }}
```
Would return:
```
ok: [localhost] => {
    "msg": [
        {
            "Name": "/Test/secure-test", 
            "Type": "SecureString", 
            "Value": "secure-test"
        }, 
        {
            "Name": "/Test/subpath/test", 
            "Type": "String", 
            "Value": "testsubpath"
        }, 
        {
            "Name": "/Test/test-string", 
            "Type": "String", 
            "Value": "test-string"
        }
    ]
}
```

You can loop through them:
```
         - name: Getting parameters in a loop
           debug: msg='key contains an item named {{item.Name }} with value {{item.Value}}'
           with_ssm:
             - '/Test'
             - 'region=ap-southeast-2'
             - 'bypath'
             - 'recursive=true'
```
would return:

```
TASK [Getting parameters in a loop] *********************************************************************
Tuesday 10 October 2017  16:22:44 +1300 (0:00:00.370)       0:00:01.035 ******* 
ok: [localhost] => (item={u'Type': u'SecureString', u'Name': u'/Test/secure-test', u'Value': u'secure-test'}) => {
    "item": {
        "Name": "/Test/secure-test", 
        "Type": "SecureString", 
        "Value": "secure-test"
    }, 
    "msg": "key contains an item named /Test/secure-test with value secure-test"
}
ok: [localhost] => (item={u'Type': u'String', u'Name': u'/Test/subpath/test', u'Value': u'testsubpath'}) => {
    "item": {
        "Name": "/Test/subpath/test", 
        "Type": "String", 
        "Value": "testsubpath"
    }, 
    "msg": "key contains an item named /Test/subpath/test with value testsubpath"
}
ok: [localhost] => (item={u'Type': u'String', u'Name': u'/Test/test-string', u'Value': u'test-string'}) => {
    "item": {
        "Name": "/Test/test-string", 
        "Type": "String", 
        "Value": "test-string"
    }, 
    "msg": "key contains an item named /Test/test-string with value test-string"
}
```

You can also ask the plugin to return short versions of parameter names (The part after the last slash, not the full path):
```
         - name: Getting parameters in a loop with names shorteting
           debug: msg='key contains an item named {{item.Name }} with value {{item.Value}}'
           with_ssm:
             - '/Test'
             - 'region=ap-southeast-2'
             - 'bypath'
             - 'shortnames'
             - 'recursive=true'
```
Would return:

```
TASK [Getting parameters in a loop with names shorteting] ***********************************************
Tuesday 10 October 2017  16:22:44 +1300 (0:00:00.371)       0:00:01.406 ******* 
ok: [localhost] => (item={u'Type': u'SecureString', u'Name': u'secure-test', u'Value': u'secure-test'}) => {
    "item": {
        "Name": "secure-test", 
        "Type": "SecureString", 
        "Value": "secure-test"
    }, 
    "msg": "key contains an item named secure-test with value secure-test"
}
ok: [localhost] => (item={u'Type': u'String', u'Name': u'test', u'Value': u'testsubpath'}) => {
    "item": {
        "Name": "test", 
        "Type": "String", 
        "Value": "testsubpath"
    }, 
    "msg": "key contains an item named test with value testsubpath"
}
ok: [localhost] => (item={u'Type': u'String', u'Name': u'test-string', u'Value': u'test-string'}) => {
    "item": {
        "Name": "test-string", 
        "Type": "String", 
        "Value": "test-string"
    }, 
    "msg": "key contains an item named test-string with value test-string"
}
```

Since you were involved in the other PR, would you be able to look at this @willthames ? 
Or does this need reviews from other maintainers?

bcoca committed:
move from with_<lookup>: to loop:

- old functionality is still available direct lookup use, the following are equivalent

  with_nested: [[1,2,3], ['a','b','c']]

  loop: "{{lookup('nested', [1,2,3], ['a','b','c'])}}"

- avoid squashing with 'loop:'
- fixed test to use new intenal attributes
- removed most of 'lookup docs' as these now reside in the plugins

bcoca committed:
config tests

also a couple of fixes to manager

bcoca committed:
add note on how to avoid service start in apt

abadger committed:
Port arg specs from type='str' to type='path'

abadger committed:
Add a hack to test whether type=path should be used in a module

abadger committed:
Fix get_exception and wildcard imports for nxos modules

abadger committed:
Remove wildcard imports from cloudstack modules.

Also add boilerplate.

abadger committed:
Update test cases
* Remove cloudstack, f5, and nxos from the wildcard import whitelist
* Remove nxos and f5 from the get_exception whitelist
* Remove some files from the pep8 legacy whitelist

abadger committed:
Remove wildcard import in test_nclu

bcoca committed:
better message for unsupported include/imports

adhoc cannot currently execute these actions

bcoca committed:
warn on bad keys in group

resmo committed:
cs_snapshot_policy: remove CloudStackException dependency

bcoca committed:
validate that existing dest is valid directory

jborean93 committed:
win_become: Added support to become a service user (#30407)

* win_become: Added support to become a service user

* fixes for linting

* changes to get local and network service working

* fixed linting issues again

* pleasing pepe

jborean93 committed:
updated changelog for win_become with service accounts

Commit:
Update ssm lookup to support getting parameters by path

Commit:
Update ssm lookup to support getting parameters by path

arcenik committed:
Fix urlparse import for Python3 (#31240)

* Fix urlparse import for Python3 in

* contrib/inventory/consul_io.py
* contrib/inventory/rudder.py
* contrib/inventory/windows_azure.py
* lib/ansible/module_utils/known_hosts.py
* lib/ansible/modules/cloud/centurylink/clc_firewall_policy.py
* lib/ansible/modules/cloud/docker/_docker.py
* lib/ansible/modules/cloud/ovirt/ovirt_disk.py
* lib/ansible/plugins/action/ce_template.py

bcoca committed:
clarified inventory_dir changes for 2.4 (#31484)

* clarified inventory_dir changes for 2.4

added example on how to restore (mostly) old behaviour for add_hosts

fixes #30901

willthames commented on lib/ansible/plugins/lookup/ssm.py:
```diff
@@ -73,6 +99,10 @@ def run(self, terms, variables, **kwargs):
                     session[key] = value
 
+                # recurse or not
+                if key == "recursive" and value.lower() == "true":
```
might want to use 

```
from ansible.module_utils.parsing.convert_bool import boolean
...

if key == "recursive" and boolean(value):
     ssm_dict['Recursive'] = True
...


for this, as people might expect to be able to use 'yes' or 'on'

mtb-xt replied:
Thanks! I've changed both *recursive* and *decrypt*

Commit:
Make ansible-test happy

ar7z1 committed:
minor updates + tests to win_firewall_rule as per jborean93 review (#29148)

* Added warning for 'force' option

* Changed 'profiles' type to list

* Changed 'interfacetypes' type to list

* Added deprecation warning and fixed doc

* updated force parameter

MikeKlebolt committed:
first path check missing forward slash between dcpath and datacenter (#31481)

* fixed folder path logic

* fixed folder path logic

* fixed folder path logic

lukas-bednar committed:
[rpm_key] Fix to import first key on the system (#31514)

* [rpm_key] Fix to import first key on the system

Fixes: #31483

* [rpm_key] removed unsafe_shell and "throwaway" underscore

* [rpm_key] adding test to add the first key on system

mkrizek committed:
yum: fix crashes installing from file/url

arnonki committed:
Use default values for object returned from aws to prevent TypeError  (#31455)

* Use default values for object returned from aws to prevent TypeError
NoneType exception fixes #31454

samdoran committed:
Remove sysctl entries when state=absent (#31486)

* Remove sysctl entry when state=absent
* Cleanup sysctl integration test syntax
* Correct grammar on error message
* Add sysctl integration test for state=absent

s-hertel committed:
aws_s3: fix upload_file's ExtraArgs - fixes #31232 (#31487)

* Module option metadata are extra arguments rather than S3 object metadata: update ExtraArgs variable.

* Remove hyphens from ExtraArgs to maintain backwards compatibility

* Map lowercase extra args to CamelCase

* Maintain backwards compatibility by guessing at content type rather than always defaulting to binary/octet-stream.

* Fix ExtraArgs for non-hyphenated options

* Simplify logic

s-hertel committed:
[cloud] Fix ec2_group when security group lacks a VPC (#31526)

rcarrillocruz committed:
Run neighbors intent tests for ios_interface only if show lldp contains netdev (#31539)

abadger committed:
Try to cleanup after the rpm_key test

nitzmahone committed:
ignore chouseknecht as azure_rm module maintainer

nitzmahone committed:
improve become/runas error messaging on bogus/missing username (#31551)

mtb-xt commented:
@ansibot  bot_status

bcoca committed:
fix paths to actually be config file relative (#31533)

* fix paths to actually be config file relative

also allow to unfrack paths for 'not cwd'
only use -i 'localhost,' for adhoc in pull

willthames commented on lib/ansible/plugins/lookup/ssm.py:
```diff
@@ -86,13 +116,34 @@ def run(self, terms, variables, **kwargs):
+                    response = client.get_parameters_by_path(NextToken=response['NextToken'], **ssm_dict)
+                    paramlist.extend(response['Parameters'])
+
+                # shortify parameter names. yes, this will return duplicate names with different values.
```
shorten, not shortify

mtb-xt replied:
Noted

bcoca committed:
more updates to plugin/config generation (#30787)

* fixed module generation

added missing lookup page
point to plugins when plugins
made modules singular
add display for verbose an debug messages
nicer templating, changed generation order for ref
corrected links
moved most of lookup docs to plugin section

* Copy edits
* Fixed typos
* Clarified wording

ganeshrn committed:
Fix rollback in junos_config (#31424)

* Fix rollback in junos_config

Fixes #30778

*  Call `load_configuration` with rollback id in case
   the id is given as input
*  Pass rollback id to `get_diff()` to fetch diff from device

* Fix unit test

Commit:
Proper words in a comment; Use ansible boolean parser; add author details

mtb-xt commented:
@willthames I've made the requested changes

rcarrillocruz committed:
Introspect the management IP on ios_ping (#31571)

On our CI we don't have external connectivity, so let's ping to
the management interface IP.
Also, ignore errors on the expected failures tests.

rcarrillocruz committed:
Remove duped authorize on ios_ping anchored tasks (#31572)

Otherwise, we get warnings.

rcarrillocruz committed:
Parameterize the ssh port on ios_user tests (#31573)

On CI we use 8022 for SSH, thus we need to explicitly pass
ansible_ssh_port to the ssh command to work.

rahushen committed:
Fix nxos_banner removal idempotence issue in N1 images (#31259)

* Fix nxos_banner removal idempotence issue in N1 images

* handle pep8 error

Commit:
New module: Add monitoring/zabbix_template module. (#25566)

* Implements zabbix_template module.

* Fixes formatting

* Fixes formatting2

* Add templates_clear

* fix some bug

* Fixes formatting

* update ansible 2.5

rahushen committed:
change ports to non well known ports and drop time_range for N1 (#31261)

pdellaert committed:
Fix #21112 : Adding support for configuring a CD-rom iso image (#28155)

* Initial CD-ROM support

* create cdrom bugfix

* Improving CDROM change detection and fixing template creation bug
Running MarkAsTemplate on an existing template will fail with an error

* Better change detection for guest ID
Should only mark a change in case it actually changes

* Adding integration tests

* Pep8 compliance fixes

* Adding CDROM support, including iso, client and none types

* Updating added release version for CDROM option

voipmeister committed:
Changed src_intf and dst_intf to list type (#31479)

ganeshrn committed:
Remove log of command executed in persistent connection (#31581)

resmo committed:
cs_snapshot_policy: fix VM not found in non-default zone

resmo committed:
tests: cs_snapshot_policy: add integration tests

resmo committed:
cs_securitygroup: remove CloudStackException dep (#31597)

- fixes pep8
- fixes docs

candlerb committed:
Add ellipsis to indicate YAML example is excerpt, not standalone

jimi-c committed:
Fix static determination for include_tasks

An incorrect removal of a conditional resulted in include_tasks falling
through to the old static detection mechanism incorrectly. This restores
the previous conditional check.

Fixes #31593

resmo committed:
cs_securitygroup_rule: remove CloudStackException dependency (#31603)

- fixes pep8
- fixes docs

schast committed:
zabbix_host: added zabbix host property (description) (#25969)

* zabbix_host: added zabbix host property (description)

* zabbix_host: fixed error E309 version_added for new option (description) should be 2.4

* zabbix_host: deleted unwanted else for update description

* lib/ansible/modules/monitoring/zabbix_host.py: increased version_added to 2.5 for option description

* lib/ansible/modules/monitoring/zabbix_host.py: allow to change the description

* lib/ansible/modules/monitoring/zabbix_host.py: added new lines back to fix pep8 issues

caphrim007 committed:
Adds new committer to Ansible core (#31544)

This is me (Tim) signing and agreeing to the committer guidelines.

jborean93 committed:
windows: removed #Requires from PS pipelines as it causes STDERR false positives (#31558)

pweldon committed:
Fix typo in command output of encrypt_string example. (#31601)

resmo committed:
cs_portforward: remove CloudStackException dep (#31610)

- fixes pep8
- fixes doc

mattclay committed:
Work around cryptography 2.1 requiring pip 8.1.2+

machacekondra committed:
ovirt_templates: Add seal parameter (#31482)

abadger committed:
Remove boundary_meter module as the hosted service it configured has been removed. (#31506)

References #29387

colin-nolan committed:
Corrects examples in the `find` module. (#31522)

cognifloyd committed:
Fix Inventory Plugin Loading (#31605)

This change makes the PluginLoader use DEFAULT_INVENTORY_PLUGIN_PATH setting.
Inventory Plugins were only being loaded the 'inventory_plugins' folder of the current directory,
as well as the ansible-provided inventory plugins (e.g. `/path/to/site-packages/ansible/plugins/inventory`).

prasadkatti committed:
[aws_s3] different fail msg depending on whether version is specified (#31598)

mattclay committed:
Update Azure tests to limit cryptography version. (#31616)

* Revert "Work around cryptography 2.1 requiring pip 8.1.2+"

This reverts commit a10a80b8a36192982829aefbba553dfdc6d9aa55.

* Update Azure tests to limit cryptography version.

gundalow committed:
botmeta for lib/ansible/plugins/filter/network.py (#31608)

resmo committed:
cs_instance_facts: fix pep8 (#31606)

resmo committed:
botmeta: maintain cloudstack module utils (#31613)

trishnaguha committed:
fix ios_interface test unsupported param (#31622)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

Akasurde committed:
pep8 fixes for vmware modules (#31537)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

trishnaguha committed:
fix nxos_vlan unexpected deprecation warning (#31624)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

mattclay committed:
Remove cryptography install from CI other test.

mattclay committed:
Multiple Parallels host support for ansible-test. (#31552)

Commit:
Create action file enos.py (#31508)

* Create action file enos_facts.py

* Update and rename enos_facts.py to enos.py

* Taking chances on Dealing with Unstable issues

* Removing blank space/ white line

Commit:
create cliconf plugin for enos - enos.py (#31509)

Commit:
create terminal plugin for enos modules (#31510)

Commit:
Create document fragment for enos modules (#31512)

Commit:
Update base.yml to add enos from lenovo (#31565)

* Update base.yml to add enos from lenovo

* Update base.yml

Commit:
Adding enos/__init__.py (#31580)

##### SUMMARY
Creating the modules for enos from Lenovo.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/modules/network/enos/__init__.py

##### ANSIBLE VERSION
ansible 2.5.0
config file = /etc/ansible/ansible.cfg
configured module search path = [u'/root/.ansible/plugins/modules',
u'/usr/share/ansible/plugins/modules']
ansible python module location =
/usr/local/lib/python2.7/dist-packages/ansible-2.5.0-py2.7.egg/ansible
executable location = /usr/local/bin/ansible
python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]

##### ADDITIONAL INFORMATION

This is an effort to create modules for enos based switches from Lenovo.

machacekondra committed:
ovirt_clusters: add alias performance_preset for memory_policy (#31523)

machacekondra committed:
ovirt_networks: Add label to network (#31517)

machacekondra committed:
ovirt_templates: Fix exlusive parameter (#31470)

machacekondra committed:
ovirt_vms: Wait for VM disks when adding VM (#31468)

Fixes: https://github.com/ansible/ansible/issues/30515

bcoca committed:
highlight dynamic parts of error msg

jctanner committed:
Only install pyvmomi if user is root (#31628)

ZubairLK committed:
vmware_vm_facts: Add error check for config.vm (#31629)

Add a simple check to see if vm.config exists before looking for
vm.config.hardware

MihaiMoldovanu committed:
etcd.py module improved . added posibility to configure from playbook… (#18580)

* etcd.py module improved . added posibility to configure from playbook etcd server to query

* Update playbooks_lookups.rst

documented etcd lookups

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Fixed errors reported by ansibot

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Copy edits.

sethp-nr committed:
wait_for: treat broken connections as "unready" (#28839)

* wait_for: treat broken connections as "unready"

We have observed the following condition while waiting for hosts:

```
Traceback (most recent call last):
  File "/var/folders/f8/23xp00654plcv2b2tcc028680000gn/T/ansible_8hxm4_/ansible_module_wait_for.py", line 585, in <module>
    main()
  File "/var/folders/f8/23xp00654plcv2b2tcc028680000gn/T/ansible_8hxm4_/ansible_module_wait_for.py", line 535, in main
    s.shutdown(socket.SHUT_RDWR)
  File "/usr/local/opt/python/Frameworks/Python.framework/Versions/2.7/lib/python2.7/socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
socket.error: [Errno 57] Socket is not connected
```

This appears to happen while the host is still starting; we believe something is
accepting our connection but immediately resetting it. In these cases, we'd
prefer to continue waiting instead of immediately failing the play.

This patch has been applied locally for some time, and we have seen no adverse
effects.

* wait_for: fixup change

We were missing an import and a space after the `#`

kairoaraujo committed:
new module: AIX rootvg backup image using mksysb (#30460)

* new module: AIX rootvg backup image using mksysb

This module is simple but very useful for AIX system
administrators. Easy to construct playbooks to generate
and manage rootvg backups using mksysb tool.

* added module_check, pep8, non-written convention

- implemented module_check;
- fixed some pep8 and non-written convention

* removed parameters as global variables and doc

Moved global variables parameters to inside main()
Better doc format for mentioned files

machacekondra committed:
ovirt: Add new api facts module (#28875)

* ovirt: Add new api facts module

* Simplify boilerplate and add Python 3 import

pilou- committed:
s3_sync: fix broken import (#31460)

bcoca committed:
update delegated var fetching for new 'loop' (#31588)

* update delegated var fetching for new 'loop'

* mixed loop/terms

tedder committed:
[cloud][docs] Document default/required 'state' in CloudFormation module (#31642)

* fix documentation to match default/required status of 'state'

* add 'choices'

sampritavh committed:
[cloud] Support changeset_name parameter on CloudFormation stack create (#31436)

mattclay committed:
Remove cryptography version limit in Azure tests.

No longer needed now that cryptography 2.1.1 has been released.

jmighion committed:
Adding a cli transport option for the bigip_command module. (#30391)

* Adding a cli transport option for the bigip_command module.

* Fixing keyerror when using other f5 modules. Adding version_added for new option in bigip_command.

* Removing local connection check because the F5 tasks can be delegated to any host that has the libraries for REST.

* Using the network_common load_provider.

* Adding unit test to cover cli transport and updating previous unit test to ensure cli was not called.

caphrim007 committed:
Adds a refactored bigip_monitor_http module. (#30998)

This patch refactors the bigip_monitor_http module to use REST
instead of SOAP. It additionally adds unit tests and current F5
code conventions.

Integration tests can be found here
* https://github.com/F5Networks/f5-ansible/blob/devel/test/integration/bigip_monitor_http.yaml
* https://github.com/F5Networks/f5-ansible/tree/devel/test/integration/targets/bigip_monitor_http/tasks

jborean93 committed:
update Windows docs (#30783)

* first round of splitting and expanding Windows docs

* Added info about WinRM and started the usage page

* more details onf usage around command/shell

* added more info

* missed a page

* added info about hotfix and ps upgrade script

* Added DSC information

* filled in FAQ and fixed up links

* changes per PR review and typo fixes found by myself

* fixed sanity check

* updated external script URLS

* Edits

* Edits

* Edits

* Copy edits.

* More copy edits.

* Added info about becoming a local service account

* some formatting changes on become and DSC indent fix

* Copy edits/rewrites

* Copy edits/rewrites

* minor updates to UAC text, typos

* minor wording updates

* minor wording tweaks/typos

* minor wording tweaks/typos

* minor tweaks/typos

* fix indent issue

* fix title underline length

* fix quote issue

ryansb committed:
Rename ssm_parameter_store module and ssm lookup to be prefixed with aws (#31662)

willthames commented:
This will need rebasing upon the rename now

wxdao committed:
Fix typo in playbooks_error_handling (#31636)

rahushen committed:
Fixes #31056 (#31057)

Lujeni committed:
pip: Use chdir directive in the venv path (fixes #25122) (#26865)

mensler committed:
Update Zabbix maintenance if it already exists to set new period, hosts etc. (#24371)

* Update Zabbix maintenance if it already exists to set new period, hosts etc.

* Whitespace

kevinburke committed:
Un-nest nested if statements (#19190)

Easier and clearer to write

if a and b:
    foo()

than

if a:
    if b:
        foo()

Lorquas committed:
updating argument validation for activation keys to match subscription-manager command line, fixes #27283 (#30236)

Akasurde committed:
Correct usage for shutil.rmtree (#31541)

* Correct usage for shutil.rmtree

Fix adds correct usage of shutil.rmtree in git module

Fixes: #31225

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

* Include archive tests so they get run

* Use new include syntax

* Cleanup syntax on git tests

- use multi-line YAML
- remove unneeded {{ }} around vars in conditionals
- remove unneeded quotes
- add task file name to task names for easier troubleshooting when things fail

* Make archive tests work for RHEL/CentOS 6

The older versions of Jinja2 in RHEL/CentOS 6 required assertion tasks using the map filter to be skipped.

The older version of git required gzip compression to be skipped on RHEL/CentOS 6.

* Account for ansible_distribution_major_version missing

KKoukiou committed:
cloud: ovirt: Add ovirt_disk_facts module (#30422)

ganeshrn committed:
Validate parse_cli filter inputs (#31293)

* Validate parse_cli filter inputs

Fixes #30517

*  Add check to validate input is of type string
*  Add check to confirm template file exist

* Update error message for parse_cli_textfsm invalid template path

* Add input validation for parse_cli_textfsm filter

caphrim007 committed:
This patch fixes a number of outstanding bugs and code convention problems. (#31618)

* documentation was not inline with other Ansible modules
* Python 3 specific imports were missing
* monitor_type is no longer required when creating a new pool; it is now the default.
* A new monitor_type choice of "single" was added for a more intuitive way to specify "a single monitor". It uses "and_list" underneath, but provides additional checks to ensure that you are specifying only a single monitor.
* host and port arguments have been deprecated for now. Please use bigip_pool_member instead.
* 'partition' field was missing from documentation.
* A note that "python 2.7 or greater is required" has been added for those who were not aware that this applies for ALL F5 modules.
* Unit tests were fixed to support the above module

sivel committed:
Port vault-keyring.py to properly load the config. Fixes #31304 (#31315)

kennethgillen committed:
Fix gender pronouns and the odd typo (#31709)

s-hertel committed:
[cloud] Document alias for aws_s3 'overwrite' option. (#31638)

s-hertel committed:
[cloud] Require src parameter when aws_s3 is used with mode 'put' (#31657)

s-hertel committed:
[cloud] Don't try to deregister an instance that isn't in the load balancer (#31660)

s-hertel committed:
[cloud] Only enforce state (running/stopped/etc) in EC2 "exact_count" when state is specified (#31648)

fix issue in ec2 module where exact count would create new instances if the instane state is stopped or terminated

alikins committed:
Vault secrets script client inc new 'keyring' client (#27669)

This adds a new type of vault-password script  (a 'client') that takes advantage of and enhances the 
multiple vault password support.

If a vault password script basename ends with the name '-client', consider it a vault password script client. 

A vault password script 'client' just means that the script will take a '--vault-id' command line arg.

The previous vault password script (as invoked by --vault-password-file pointing to an executable) takes
no args and returns the password on stdout. But it doesnt know anything about --vault-id or multiple vault
passwords.

The new 'protocol' of the vault password script takes a cli arg ('--vault-id') so that it can lookup that specific
vault-id and return it's password.

Since existing vault password scripts don't know the new 'protocol', a way to distinguish password scripts
that do understand the protocol was needed.  The convention now is to consider password scripts that are
named like 'something-client.py' (and executable) to be vault password client scripts.

The new client scripts get invoked with the '--vault-id' they were requested for. An example:

     ansible-playbook --vault-id my_vault_id@contrib/vault/vault-keyring-client.py some_playbook.yml

That will cause the 'contrib/vault/vault-keyring-client.py' script to be invoked as:

     contrib/vault/vault-keyring-client.py --vault-id my_vault_id

The previous vault-keyring.py password script was extended to become vault-keyring-client.py. It uses
the python 'keyring' module to request secrets from various backends. The plain 'vault-keyring.py' script
would determine which key id and keyring name to use based on values that had to be set in ansible.cfg.
So it was also limited to one keyring name.

The new vault-keyring-client.py will request the secret for the vault id provided via the '--vault-id' option.
The script can be used without config and can be used for multiple keyring ids (and keyrings).

On success, a vault password client script will print the password to stdout and exit with a return code of 0.
If the 'client' script can't find a secret for the --vault-id, the script will exit with return code of 2 and print an error to stderr.

bcoca committed:
additional configmanager tests

left placeholders for more

mattclay committed:
Fix Debian packaging Dockerfile and docs.

resmo committed:
cs_instance: remove CloudStackException dep (#31724)

- fixes dep
- fix docs

viper233 commented:
I have a problem with lookups within the AWS environment as it seems impossible to pass STS tokens for cross account access. Seeing aws_ssm_parameter_store exists, wouldn't it be better to have aws_ssm_parameter_store_facts?

caphrim007 committed:
Fixes various gtm pool issues (#31728)

Various formatting related fixes. Also fixed an idempotency problem
with the 'disabled' state

caphrim007 committed:
Fixes documentation related bugs (#31730)

New conventions for ansible warrant fixes to accomodate those
in bigip_partition.

This patch also includes an import fix that can raise an error when
Ansible unit tests run

caphrim007 committed:
Adds various provision fixes (#31731)

* vcmp provisioning support
* documentation fixes
* fixes for python3 causing an exception

caphrim007 committed:
Refactors bigip_selfip (#31732)

In this refactor we moved to the most recent coding standards for
both F5 and Ansible. Many bugs were fixed and some features were
also added (such as ipv6 support).

caphrim007 committed:
Fixes an fqdn error (#31745)

the pool module was not correctly interpreting the partitions
in the monitors list vs the partition at the task level. This patch
fixes that.

mtb-xt commented:
@willthames I've screwed up rebasing this PR, so I opened another one - https://github.com/ansible/ansible/pull/31752 . Sorry about that :/
